core/src/test/java/io/questdb/test/griffin/ParallelGroupByTest.java

```diff
@@ -0,0 +1,906 @@
+                "SELECT key1, key2, avg + sum from (" +
+                        "  SELECT key1, key2, avg(value), sum(colTop) FROM tab" +
+                        ")",
+                "key1\tkey2\tcolumn\n" +
+                        "k1\tk1\t20103.5\n" +
+                        "k2\tk2\t20184.5\n" +
+                        "k3\tk3\t20265.5\n" +
+                        "k4\tk4\t20346.5\n" +
+                        "k0\tk0\t20427.5\n"
+        );
+    }
+
+    @Test
+    public void testParallelMultiKeyedGroupByWithFilter() throws Exception {
+        testParallelMultiKeyedGroupBy(
+                "SELECT key1, key2, avg(value), sum(colTop) FROM tab WHERE value < 80",
+                "key1\tkey2\tavg\tsum\n" +
+                        "k1\tk1\t45.31818181818182\t381.0\n" +
+                        "k2\tk2\t46.31818181818182\t387.0\n" +
+                        "k3\tk3\t47.31818181818182\t393.0\n" +
+                        "k4\tk4\t48.31818181818182\t399.0\n" +
+                        "k0\tk0\t46.25\t325.0\n"
+        );
+    }
+
+    @Test
+    public void testParallelMultiKeyedGroupByWithLimit() throws Exception {
+        testParallelMultiKeyedGroupBy(
+                "SELECT key1, key2, avg(value), sum(colTop) FROM tab LIMIT 3",
+                "key1\tkey2\tavg\tsum\n" +
+                        "k1\tk1\t223.5\t19880.0\n" +
+                        "k2\tk2\t224.5\t19960.0\n" +
+                        "k3\tk3\t225.5\t20040.0\n",
+                "SELECT key1, key2, avg(value), sum(colTop) FROM tab LIMIT -3",
+                "key1\tkey2\tavg\tsum\n" +
+                        "k3\tk3\t225.5\t20040.0\n" +
+                        "k4\tk4\t226.5\t20120.0\n" +
+                        "k0\tk0\t227.5\t20200.0\n"
+        );
+    }
+
+    @Test
+    public void testParallelMultiKeyedGroupByWithNestedFilter() throws Exception {
+        testParallelMultiKeyedGroupBy(
+                "SELECT avg(v), sum(ct), k1, k2 " +
+                        "FROM (SELECT value v, colTop ct, key2 k2, key1 k1 FROM tab WHERE value < 80)",
+                "avg\tsum\tk1\tk2\n" +
+                        "45.31818181818182\t381.0\tk1\tk1\n" +
+                        "46.31818181818182\t387.0\tk2\tk2\n" +
+                        "47.31818181818182\t393.0\tk3\tk3\n" +
+                        "48.31818181818182\t399.0\tk4\tk4\n" +
+                        "46.25\t325.0\tk0\tk0\n"
+        );
+    }
+
+    @Test
+    public void testParallelNonKeyedGroupBy() throws Exception {
+        testParallelNonKeyedGroupBy(
+                "SELECT vwap(price, quantity), sum(colTop) FROM tab",
+                "vwap\tsum\n" +
+                        "289.3066666666667\t100200.0\n"
+        );
+    }
+
+    @Test
+    public void testParallelNonKeyedGroupByConcurrent() throws Exception {
+        final int numOfThreads = 8;
+        final int numOfIterations = 50;
+        final String query = "SELECT avg(value), sum(colTop) FROM tab";
```

Comment:
Does this query use the new parallel execution? I don't see a filter here.

Reply:
Can we add some fuzziness into this, e.g. a fuzzy filter that sometimes returns all sometimes no rows? Or random query and compare that parallel and non-parallel execution returns the same results?

Reply:
> Does this query use the new parallel execution? I don't see a filter here.

Parallel execution doesn't need a filter. It kicks in on any suitable GROUP BY. We don't just filter the data - that's optional. The main action is aggregating the data in parallel.

Reply:
> Or random query and compare that parallel and non-parallel execution returns the same results?

This would need a SQL fuzzer. Let me think of it.

Reply:
Looks like this suggestion is the only pending one. I'm looking at other OSS projects trying to find a SQL fuzzer we could reuse. If the search fails, I'll try writing a simple fuzzer for GROUP BY queries.